_posts/2020-07-12-ray-tune.md

```diff
@@ -0,0 +1,442 @@
+
+[ray-trainable]:https://medium.com/rapids-ai/30x-faster-hyperparameter-search-with-raytune-and-rapids-403013fbefc5
+
+In addition, I have benchmarked Dask-ML's model selection module to see how the
+time-to-solution is affected by the number of Dask workers in "[Better and
+faster hyperparameter optimization with Dask][db-bf]." That is, how does the
+time to reach a particular accuracy scale with the number of workers $P$? At
+first, it'll scale like $1/P$ but with large number of workers the serial
+portion will dictate time to solution according to [Amdahl's Law]. Briefly, I
+found Dask-ML's `HyperbandSearchCV` speedup started to saturate around 24
+workers for a particular search.
+
+[^bohb-parallel]:In Section 4.2 of [their paper](http://proceedings.mlr.press/v80/falkner18a/falkner18a.pdf).
+
+[db-bf]:https://blog.dask.org/2019/09/30/dask-hyperparam-opt
+
+[Amdahl's Law]:https://en.wikipedia.org/wiki/Amdahl%27s_law
+[ddo]:https://docs.dask.org/en/latest/setup.html
+[plg]:https://lightgbm.readthedocs.io/en/latest/Parallel-Learning-Guide.html
+
+### Speed
+
+*Dask-ML and Ray are much faster than Scikit-Learn.*
+
+Ray's tune-sklearn runs some benchmarks in [the introduction][1] with the
+`GridSearchCV` class found Scikit-Learn and Dask-ML. A more fair benchmark
+would be use Dask-ML's `HyperbandSearchCV` because it is almost the same as the
+algorithm in Ray's tune-sklearn. To be specific, I'm interested in comparing
+these methods:
+
+* Scikit-Learn's `RandomizedSearchCV`. This is a popular implementation, one
+  that I've bootstrapped myself with a custom model.
+* Dask-ML's `HyperbandSearchCV`. This is an early stopping technique for
+  `RandomizedSearchCV`.
+* Ray tune-sklearn's `TuneSearchCV`. This is a slightly different early
+  stopping technique than `HyperbandSearchCV`'s.
+
+Each of these will be configured to perform the same tasks, sample 100
+parameters and train for no longer than 100 "epochs" or passes through the
+data. All of these estimators will be set up as their respective documentation
+suggests;[^random-search] the complete setup is in [the
+appendix](#appendix).
+
+Each search will use 8 workers with a single cross validation split. I've also
+setup the model so every `partial_fit` call processes 60,000 examples in 1.2
+seconds. Here's how long each library takes to complete this search:
+
+<img src="/images/2020-model-selection/n_workers=8.png" width="450px"
+ />
+
+
+Clearly, Ray and Dask-ML offer similar performance for 8 workers when compared
+with Scikit-Learn. Even relatively, Dask-ML is only 15% slower than Ray with 8
+workers.  Ray might perform better with more workers because they implement an
+asynchronous variant of Dask-ML's Hyperband; however, Dask-ML can scale up to 24
+workers for the search in "[Better and faster hyperparameter optimization with
+Dask][db-bf]." That's the number of GPUs on 4 nodes of the [Summit
+supercomputer][summit]!
```

Comment:
```suggestion
with Scikit-Learn. However to Ray's credit we do find that the Dask-ML HyperbandSearchCV
algorithm is 15% slower than Ray with 8 workers.  
We suspect that Ray Tune performs better in this case because they implement an
asynchronous variant of Dask-ML's Hyperband.
```

Unpacking this suggestion a bit I think that the phrasing "Dask is only 15% slower" sounds a bit defensive.  It sounds like we're coming from a position of an underdog.  Instead, I recommend a more gracious tone, which I've tried to produce here with "To Ray's credit ..."

I didn't understand the excitement around using four nodes of Summit.  This sounds like a small number of nodes to me.  I may not understand the scaling challenges in this space though.  

If you wanted to add more credibility on the Dask side you might say things like "We haven't yet investigated this slowdown.  It's worth noting that we didn't do any optimizing or tuning of the Dask-ML codebase when running this benchmark.  This is the code as it has been for the last year."  You might also refer to the [Biased Benchmarks](http://matthewrocklin.com/blog/work/2017/03/09/biased-benchmarks) post.

Reply:
Oh, I see that there is a reference to this post just below already.